File: app/src/main/java/com/example/benchmarks/ui/benchmark/BenchmarksViewModel.java
package com.example.benchmarks.ui.benchmark;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.benchmarks.models.Benchmark;
import com.example.benchmarks.models.BenchmarkItem;
import com.example.benchmarks.utils.Pair;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.disposables.Disposable;
import io.reactivex.rxjava3.schedulers.Schedulers;

public class BenchmarksViewModel extends ViewModel {

    private final MutableLiveData<List<BenchmarkItem>> itemsLiveData = new MutableLiveData<>();
    private final MutableLiveData<Integer> testSizeLiveData = new MutableLiveData<>();
    private final MutableLiveData<Boolean> calculationStartLiveData = new MutableLiveData<>(false);
    private final Benchmark benchmark;
    private Disposable disposable = Disposable.disposed();

    public BenchmarksViewModel(Benchmark benchmark) {
        this.benchmark = benchmark;
    }

    public void onCreate(){
        itemsLiveData.setValue(benchmark.createBenchmarkList(false));
    }

    public static Pair<Boolean, Integer> isNumberCorrect(String number) {
        try {
            int temp = Integer.parseInt(number);
            return new Pair<>(temp > 0, temp);
        } catch (NumberFormatException exception) {
            return new Pair<>(false, 0);
        }
    }

    public void onButtonToggle() {
        if (disposable.isDisposed()) {
            onStartProcess();
        } else {
            onStopProcess();
        }
    }

    private void onStartProcess() {
        final List<BenchmarkItem> items = benchmark.createBenchmarkList(true);
        calculationStartLiveData.setValue(true);
        final int testSize = testSizeLiveData.getValue() == null ? 0 : testSizeLiveData.getValue();

        disposable = Flowable.fromIterable(items)
                .map(benchmarkItem -> {
                    final long time = benchmark.measureTime(testSize, benchmarkItem);
                    return new Pair<>(items.indexOf(benchmarkItem), benchmarkItem.updateBenchmarkItem(time));
                })
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnComplete(this::onStopProcess)
                .subscribe(this::recreateItemsList);
    }

    private void onStopProcess() {
        calculationStartLiveData.setValue(false);
        if (!disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    private void recreateItemsList(Pair<Integer, BenchmarkItem> benchmarkResult) {
        List<BenchmarkItem> list = itemsLiveData.getValue();
        if (list != null) {
            List<BenchmarkItem> newList = new ArrayList<>(list);
            newList.set(benchmarkResult.first, benchmarkResult.second);
            itemsLiveData.setValue(newList);
        }
    }

    public int getCountOfSpans(){
        return benchmark.getSpansCount();
    }

    public LiveData<List<BenchmarkItem>> getItemsLiveData() {
        return itemsLiveData;
    }

    public void setSizeCollectionLiveData(Integer size) {
        testSizeLiveData.setValue(size);
    }

    public LiveData<Boolean> getCalculationStartLiveData() {
        return calculationStartLiveData;
    }

}

File: app/src/main/java/com/example/benchmarks/ui/benchmark/BenchmarksViewModel.kt
package com.example.benchmarks.ui.benchmark

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.benchmarks.models.benchmark.Benchmark
import com.example.benchmarks.models.benchmark.BenchmarkItem
import com.example.benchmarks.utils.DispatchersHolder
import com.example.benchmarks.utils.Pair
import kotlinx.coroutines.*

class BenchmarksViewModel(
    private val benchmark: Benchmark,
    private val dispatchers: DispatchersHolder
) : ViewModel() {

    private val itemsLiveData = MutableLiveData<List<BenchmarkItem>>()
    private val testSizeLiveData = MutableLiveData<Int>()
    private val calculationStartLiveData = MutableLiveData(false)
    private var job: Job? = null

    companion object {
        @JvmStatic
        fun isNumberCorrect(number: String): Pair<Boolean, Int>? {
            return try {
                val temp = number.toInt()
                Pair(temp > 0, temp)
            } catch (exception: NumberFormatException) {
                Pair(false, 0)
            }
        }
    }

    fun onCreate() {
        itemsLiveData.value = benchmark.createBenchmarkList(false)
    }

    fun onButtonToggle() {
        if (job?.isActive == true) {
            onStopProcess()
        } else {
            onStartProcess()
        }
    }

    private fun onStartProcess() {
        val items = benchmark.createBenchmarkList(true)
        itemsLiveData.postValue(items)
        calculationStartLiveData.value = true
        val testSize = testSizeLiveData.value ?: 0

        job = viewModelScope.launch(dispatchers.getIO()) {
            val deferredList: List<Deferred<BenchmarkItem>> = items.map { item ->
                async {
                    val time = benchmark.measureTime(testSize, item)
                    item.updateBenchmarkItem(time)
                }
            }
            val resultList = deferredList.awaitAll()
            itemsLiveData.postValue(resultList)
            calculationStartLiveData.postValue(false)
        }
    }

    private fun onStopProcess() {
        if (job?.isActive == true){
            job?.cancel()
            calculationStartLiveData.value = false
        }
    }

    fun getCountOfSpans(): Int = benchmark.getSpansCount()

    fun setSizeCollectionLiveData(size: Int) {
        testSizeLiveData.value = size
    }

    fun getItemsLiveData(): LiveData<List<BenchmarkItem>> = itemsLiveData

    fun getCalculationStartLiveData(): LiveData<Boolean> = calculationStartLiveData
}
